### PR TITLE
biomes functionallity and encounter modifier for road-generation

### PR DESCRIPTION
**Pull request by GreenUserBlue**

### CODE REVIEW

File: core/src/com/fourinachamber/fortyfive/map/detailMap/MapEvent.kt [hunk 1]
```diff
@@ -18,13 +16,18 @@ object MapEventFactory {
...
         "NPCMapEvent" to { NPCMapEvent(it.get<String>("npc")) },
-        "ShopMapEvent" to { it ->
+        "ShopMapEvent" to {
             ShopMapEvent(
                 it.get<String>("type"),
-                it.get<String>("biome"),
                 it.get<String>("person"),
                 it.get<Long?>("seed") ?: (Math.random() * 1000).toLong(),
-                it.get<List<OnjInt>>("boughtIndices").map{it2->it2.value.toInt()}.toMutableSet(),
+                it.get<List<OnjInt>>("boughtIndices").map { it2 -> it2.value.toInt() }.toMutableSet(),
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/detailMap/MapEvent.kt:24]: when nesting lambdas like this, please give the parameters proper names, not `it2`

File: core/src/com/fourinachamber/fortyfive/map/detailMap/MapEvent.kt [hunk 2]
```diff
@@ -18,13 +16,18 @@ object MapEventFactory {
...
+            )
+        },
+        "ChooseCardMapEvent" to {
+            ChooseCardMapEvent(
+                it.get<OnjArray>("types").value.map { t -> (t as OnjString).value },
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/detailMap/MapEvent.kt:29]: please don't use single letter variable names

File: core/src/com/fourinachamber/fortyfive/map/detailMap/SeededMapGenerator.kt [hunk 3]
```diff
@@ -104,12 +114,36 @@ class SeededMapGenerator(
...
+                .toMutableList(), rnd)
+        repeat(restrictions.nbrOfEncounters[max(nbrOfEncounterIndex, 0)].first) {
+            if (encounterProps.isEmpty()) return res
+            val encIndex = RandomCardSelection.getRandomIndex(encounterProps.map { it.second.toFloat() }
+                .toMutableList(), rnd)
+            try {
+                res.add((encounterProps[encIndex].first))
+            } catch (e: Exception) {
+                throw Exception("Unknown encounter Modifier: ${encounterProps[encIndex]}. Possibilities")
+            }
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/detailMap/SeededMapGenerator.kt:139]: why catch an exception and then immedeately rethrow another? It's often better to let the original exception through, in order to preserve the stack trace

File: core/src/com/fourinachamber/fortyfive/map/events/RandomCardSelection.kt [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    /**
+     * @param cards the cards to take as source (can have doubles)
+     * @param changeNames the names of the types you want to apply to
+     * @param cardsMaxOnce if true, then all chosen cards will be removed from the remaining selection and can therefore be there only once
+     * @param nbrOfCards how many cards you want. However, there can be fewer cards remaining after applying the effects
+     * @throws Exception if a type is not known
+     */
+    fun <T> getRandomCards(
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/events/RandomCardSelection.kt:75]: I struggle to understand what this function does, maybe the change the name of the function/parameters to be more clear and/or update the doc-comment

File: core/src/com/fourinachamber/fortyfive/map/events/RandomCardSelection.kt [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        rnd: Random,
+        biome: String,
+        occasion: String,
+    ): List<T> {
+        if (nbrOfCards >= cards.size && cardsMaxOnce) return cards
+        val (tempCards, tempChances) = getCardsWithChances(cards.toMutableList(), changeNames, biome, occasion)
+        return getCardsFromChances(nbrOfCards, tempCards, tempChances, rnd, cardsMaxOnce)
+    }
+
+    fun <T> getCardsFromChances(
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/events/RandomCardSelection.kt:89]: What is the difference between `getCardsFromChances` and `getCardsWithChances`? I don't really understand what any of these functions do

File: core/src/com/fourinachamber/fortyfive/map/events/chooseCard/ChooseCardScreenController.kt [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    private val cardsFilePath = onj.get<String>("cardsFile")
+    private val leaveButtonName = onj.get<String>("leaveButtonName")
+    private val cardsParentName = onj.get<String>("cardsParentName")
+    private val addToDeckWidgetName = onj.get<String>("addToDeckWidgetName")
+    private val addToBackpackWidgetName = onj.get<String>("addToBackpackWidgetName")
+    private var context: ChooseCardMapEvent? = null
+    private lateinit var addToDeckWidget: CustomImageActor
+    private lateinit var addToBackpackWidget: CustomImageActor
+    private var screen: OnjScreen? = null
+    override fun init(onjScreen: OnjScreen, context: Any?) {
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/events/chooseCard/ChooseCardScreenController.kt:37]: blank line missing

File: core/src/com/fourinachamber/fortyfive/map/events/chooseCard/ChooseCardScreenController.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                parent,
+                screen
+            )!! as CustomImageActor
+            curActor.name = curCard.name
+            curActor.programmedDrawable = TextureRegionDrawable(curCard.actor.pixmapTextureRegion)
+        }
+    }
+
+    private fun addListener(screen: OnjScreen) {
+        (screen.namedActorOrError(leaveButtonName) as CustomLabel).addListener(object : ClickListener() {
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/events/chooseCard/ChooseCardScreenController.kt:102]: can't you just use a normal click-listener like `onButtonClick` here?

File: core/src/com/fourinachamber/fortyfive/map/events/shop/ShopCardsHandler.kt [hunk 8]
```diff
@@ -24,30 +23,42 @@ class ShopCardsHandler(
...
+        } catch (e: Exception) {
+            RandomCardSelection.getRandomCards(
+                _allCards,
+                listOf(defaultType),
+                true,
+                nbrOfItems,
+                rnd,
+                MapManager.currentDetailMap.biome,
+                "shop"
+            )
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/map/events/shop/ShopCardsHandler.kt:57]: don't use exceptions as control flow

File: core/src/com/fourinachamber/fortyfive/screen/general/CustomActors.kt [hunk 9]
```diff
@@ -339,15 +339,22 @@ open class CustomImageActor @AllThreadsAllowed constructor(
...
         set(value) {
             if (field != value) loadedDrawable = null
             field = value
         }
-
     protected var loadedDrawable: Drawable? = null
         private set
 
+    /**
+     * overrites and ignores the background handle and the loaded drawable
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/screen/general/CustomActors.kt:354]: typo: "overrides"

File: core/src/com/fourinachamber/fortyfive/screen/general/DragAndDropBehaviours.kt [hunk 10]
```diff
@@ -172,3 +197,37 @@ open class ExecutionPayload { //TODO maybe interface, but idk how to set "tasks"
 typealias DragBehaviourCreator = (DragAndDrop, Actor, OnjNamedObject) -> DragBehaviour
 typealias DropBehaviourCreator = (DragAndDrop, Actor, OnjNamedObject) -> DropBehaviour
 
+class OnClickToCenter(private val src: CenteredDragSource) : ClickListener() {
+
+    var realStart = false
+    private var startOffset: Vector2? = Vector2()
+    override fun touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Boolean {
+        if (button == 0) {
+            if (src.fakeStart(event, x, y, pointer)) {
```
Comment (blueUserRed) [inline core/src/com/fourinachamber/fortyfive/screen/general/DragAndDropBehaviours.kt:206]: could be refactored as a guard clause

**Review by blueUserRed — APPROVED** [review]
good pr, I only found some relativley minor issues regarding code style

File: assets/maps/events/shops.onj [hunk 11]
```diff
@@ -2,106 +2,14 @@ import "imports/colors.onj" as color;
...
-
+var worldWidth = 1600.0;
+var worldHeight = 900.0;
 
 people: [
     {
         name: "traveling_merchant",
         displayName: "Traveling Merchant",
-        defaultShopParameter: "no_biome",
+        defaultShopParameter: "wasteland",
```
[the remark this answers is not in the document]
Comment (GreenUserBlue) [reply assets/maps/events/shops.onj:12]: this was just a bad name, the default shop parameter is actually the type, and has nothing to do with the biome, but i named a type after a biome, which is stupid

File: assets/screens/choose_card_screen.onj [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            ],
+        },
+        $Label {
+            text: "Drag it to add it to your Deck or Backpack.",
+            font: "red_wing_cm",
+            color: color.forty_white,
+            fontScale: 0.7,
+            align: "center",
+        },
+        //        $Label {   //this one not used because of phillip beeing indesicive
```
[the remark this answers is not in the document]
Comment (GreenUserBlue) [reply assets/screens/choose_card_screen.onj:225]: ik, that was on purpose

**Pull request merged**